rust/chains/hyperlane-cosmos/src/mailbox.rs

```diff
@@ -0,0 +1,378 @@
+use base64::Engine;
+use std::fmt::{Debug, Formatter};
+use std::io::Cursor;
+use std::num::NonZeroU64;
+use std::ops::RangeInclusive;
+
+use crate::grpc::{WasmGrpcProvider, WasmProvider};
+use crate::payloads::mailbox::{
+    GeneralMailboxQuery, ProcessMessageRequest, ProcessMessageRequestInner,
+};
+use crate::payloads::{general, mailbox};
+use crate::rpc::{CosmosWasmIndexer, WasmIndexer};
+use crate::CosmosProvider;
+use crate::{signers::Signer, verify, ConnectionConf};
+use async_trait::async_trait;
+
+use cosmrs::proto::cosmos::base::abci::v1beta1::TxResponse;
+use cosmrs::proto::cosmos::tx::v1beta1::SimulateResponse;
+use cosmrs::tendermint::abci::EventAttribute;
+
+use crate::binary::h256_to_h512;
+use hyperlane_core::{
+    utils::fmt_bytes, ChainResult, HyperlaneChain, HyperlaneContract, HyperlaneDomain,
+    HyperlaneMessage, HyperlaneProvider, Indexer, LogMeta, Mailbox, TxCostEstimate, TxOutcome,
+    H256, U256,
+};
+use hyperlane_core::{ContractLocator, Decode, RawHyperlaneMessage, SequenceIndexer};
+use tracing::{instrument, warn};
+
+/// A reference to a Mailbox contract on some Cosmos chain
+pub struct CosmosMailbox {
+    _conf: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    signer: Signer,
+    provider: Box<WasmGrpcProvider>,
+}
+
+impl CosmosMailbox {
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(conf: ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider: WasmGrpcProvider =
+            WasmGrpcProvider::new(conf.clone(), locator.clone(), signer.clone());
+
+        Self {
+            _conf: conf,
+            domain: locator.domain.clone(),
+            address: locator.address,
+            signer,
+            provider: Box::new(provider),
+        }
+    }
+}
+
+impl HyperlaneContract for CosmosMailbox {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosMailbox {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(self.domain.clone()))
+    }
+}
+
+impl Debug for CosmosMailbox {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        // Debug::fmt(&(self as &dyn HyperlaneContract), f)
+        todo!()
+    }
+}
+
+#[async_trait]
+impl Mailbox for CosmosMailbox {
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn count(&self, lag: Option<NonZeroU64>) -> ChainResult<u32> {
+        let payload = mailbox::NonceRequest {
+            nonce: general::EmptyStruct {},
+        };
+
+        let data = self
+            .provider
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, lag)
+            .await;
+
+        if let Err(e) = data {
+            warn!("error: {:?}", e);
```

Comment:
move this to a `?` above and remove the warn!, we instrument anyways so will see a log